docs/design/2019-11-14-tidb-builtin-diagnostics-zh_CN.md

```diff
@@ -0,0 +1,678 @@
+| |   Metrics   | |             |        |
+| +-------------+ |             |        |
+| +-------------+ |             |        |
+| |   Disk IO   | +---L1:gRPC-->+        |
+| +-------------+ |             |        |
+| +-------------+ |             |  TiDB  |
+| |  Network IO | |             |        |
+| +-------------+ |             |        |
+| +-------------+ |             |        |
+| |   Hardware  | +---L1:HTTP-->+        |
+| +-------------+ |             |        |
+| +-------------+ |             |        |
+| | System Info | |             |        |
+| +-------------+ |             |        |
++-----------------+             +---+----+
+                                    | 
+                   +---infoschema---+ 
+                   |                  
+                   v                  
++-L4---------------+---------------------+
+|                                        |
+|          Diagnosis Framework           |
+|                                        |
+| +---------+ +---------+  +---------+   |
+| | rule1   | |  rule2  |  |  rule3  |   |
+| +---------+ +---------+  +---------+   |
++----------------------------------------+
+```
+
+### 系统信息收集
+
+TiDB/TiKV/PD 三个组件都需要实现系统信息采集模块，其中 TiDB/PD 使用 Golang 实现并复用逻辑，TiKV 需要使用 Rust 单独实现。
+
+#### 节点硬件信息
+
+各个节点需要获取的硬件信息包括：
+
+- CPU 信息：物理核心数、逻辑核心数量、NUMA 信息、CPU 频率、CPU 供应商、L1/L2/L3 缓存大小
+- 网卡信息：网卡设备名、网卡是否启用、生产厂商、型号、带宽、驱动版本、接口队列数（可选）
+- 磁盘信息：磁盘名、磁盘容量、磁盘使用量、磁盘分区、挂载信息
+- USB 设备列表
+- 内存信息
+
+#### 节点系统信息
+
+各个节点需要获取的系统信息包括：
+
+- CPU 使用率、1/5/15 分钟负载
+- 内存：Total/Free/Available/Buffers/Cached/Active/Inactive/Swap
+- 磁盘 IO：
+    - tps: 该设备每秒的传输次数
+    - rrqm/s: 每秒这个设备相关的读取请求有多少被 Merge
+    - wrqm/s: 每秒这个设备相关的写入请求有多少被 Merge
+    - r/s: 每秒从设备读取的数据量
+    - w/s: 每秒从设备写入的数据量
+    - rsec/s: 每秒读取的扇区数
+    - wsec/s: 每秒写取的扇区数
+    - avgrq-sz: 平均请求扇区的大小
+    - avgqu-sz: 是平均请求队列的长度
+    - await: 每一个IO请求的处理的平均时间（单位是微秒毫秒）
+    - svctm: 表示平均每次设备I/O操作的服务时间（以毫秒为单位）
+    - %util: 在统计时间内所有处理IO时间，除以总共统计时间
+- 网络 IO
+    - IFACE：LAN接口
+    - rxpck/s：每秒钟接收的数据包
+    - txpck/s：每秒钟发送的数据包
+    - rxbyt/s：每秒钟接收的字节数
+    - txbyt/s：每秒钟发送的字节数
+    - rxcmp/s：每秒钟接收的压缩数据包
+    - txcmp/s：每秒钟发送的压缩数据包
+    - rxmcst/s：每秒钟接收的多播数据包
+- 常用的系统配置：sysctl -a
+
+#### 节点配置信息
+#### 节点日志信息
+所有节点都包含当前节点的生效配置，不需要额外的步骤既可拿到配置信息。
+
+#### 节点日志查询
+- 日志分布在各个节点，需要登陆到每一个节点使用关键字进行搜索
+TiDB 集群部署过程中没有部署额外的日志收集组件，TiDB/TiKV/PD 产生的日志都保存在各自的节点上，在日志检索中主要遇到以下问题：
+
+- 日志分布在各个节点，需要单独登陆到每一个节点使用关键字进行搜索
+要解决以上问题，有以下两种解决方案：
+- 没有简单的方式对多个节点的日志按照时间排序整合到同一个文件
+
+    - 优势：统一的日志管理，日志可以长时间保存，并易于检索，并且多个组件的日志可以按照时间排序归并
+    - 劣势：增加集群运维难度，第三方组件不容易与 TiDB SQL 引擎集成；日志收集工具会收集全量日志，收集过程占用各个系统资源（磁盘 IO、网络 IO）
+- 引入第三方日志收集组件对所有节点的日志进行收集
+    - 优势：不引入第三方组件，谓词下推后只返回过滤后的日志，能轻易的与 TiDB SQL 进行集成，并能复用 SQL 引擎的过滤、聚合等
+    - 劣势：增加集群运维难度，第三方组件不容易与 TiDB 内部 SQL 集成；日志收集工具会收集全量日志，收集过程占用各个系统资源（磁盘 IO、网络 IO）
+- 各个节点提供日志服务，TiDB 通过各个节点的接口将谓词下推到日志检索接口，直接对各个节点返回的日志进行归并
+    - 优势：不引入三方组件，谓词下推后只返回过滤后的日志，能轻易的与 TiDB SQL 进行集成，并能复用 SQL 引擎的过滤、聚合等
+    - 劣势：如果节点日志删除后，不能检索到对应日志
```

Comment:
And there's a problem about stability of nodes. Because logs can be very huge, so it's consuming to process them.
But if it's a third-party component, then the resources can be separated from TiDB servers.

Reply:
If the predicate contains time range, just log files in this range will be searched and will limit the returned log items count, so I think it's acceptable.

Reply:
So you will limit the SQL user requests? e.g. there must be a `WHERE`, `LIMIT`, no `ORDER BY`?

Reply:
It will sort the log items by the time default. It will sort by user define expression if the `ORDER BY` clause existed (the sort executor executed in TiDB instead of remote node).

Reply:
If there is no `WHERE` clause and no `LIMIT` clause, we will return limited log items.